File: src/lib/desktop-update.ts
import { invoke } from "@tauri-apps/api/core"
import { DESKTOP_VERSION } from "@/lib/build-info"

export const DESKTOP_UPDATE_DIALOG_EVENT = "trading-review-wiki.open-update-dialog"
export const DESKTOP_UPDATE_DOWNLOAD_URL =
  import.meta.env.VITE_DESKTOP_UPDATE_DOWNLOAD_URL || "https://api.huanxing.ai/tradereview-client"
export const DESKTOP_UPDATE_MANIFEST_URL =
  import.meta.env.VITE_DESKTOP_UPDATE_MANIFEST_URL ||
  "https://ai.fengchiyun.com/downloads/huanxingtradereview/latest.json"

export interface DesktopUpdateAsset {
  label?: string
  platform?: string
  fileName?: string
  size?: number
  sha256?: string
  url?: string
  versionedUrl?: string
  sourceUrl?: string
  baiduPanUrl?: string
  baiduPanCode?: string
  quarkPanUrl?: string
  quarkPanCode?: string
}

export interface DesktopUpdateManifest {
  repository?: string
  version?: string
  semver?: string
  publishedAt?: string
  sourceUrl?: string
  updatedAt?: string
  assets?: Record<string, DesktopUpdateAsset>
}

export interface DesktopUpdateManifestFetchResult {
  ok: boolean
  manifestUrl: string
  manifest?: DesktopUpdateManifest | null
  error?: string | null
  checkedAtMs: number
}

export interface DesktopUpdateCheckResult extends DesktopUpdateManifestFetchResult {
  updateAvailable: boolean
  currentVersion: string
  latestVersion?: string
  downloadUrl: string
  sourceUrl?: string
}

export interface DesktopInstallUpdateResult {
  ok: boolean
  manifestUrl: string
  asset?: DesktopUpdateAsset | null
  filePath?: string | null
  bytesDownloaded: number
  bytesTotal?: number | null
  launched: boolean
  error?: string | null
}

export interface DesktopInstallUpdateProgress {
  stage: "starting" | "downloading" | "verifying" | "launching" | "complete" | "error" | string
  bytesDownloaded: number
  bytesTotal?: number | null
  percent?: number | null
  fileName?: string | null
  message?: string | null
}

interface ParsedSemver {
  major: number
  minor: number
  patch: number
  prerelease: string[]
}

const SEMVER_RE = /^v?(\d+)\.(\d+)\.(\d+)(?:-([0-9A-Za-z.-]+))?(?:\+[0-9A-Za-z.-]+)?$/

function parseIntegerPart(value: string): number | null {
  if (value.length > 1 && value.startsWith("0")) return null
  const parsed = Number(value)
  return Number.isSafeInteger(parsed) && parsed >= 0 ? parsed : null
}

function parseDesktopSemver(raw: string | null | undefined): ParsedSemver | null {
  const value = String(raw ?? "").trim()
  const match = SEMVER_RE.exec(value)
  if (!match) return null

  const major = parseIntegerPart(match[1])
  const minor = parseIntegerPart(match[2])
  const patch = parseIntegerPart(match[3])
  if (major == null || minor == null || patch == null) return null

  const prerelease = match[4]?.split(".") ?? []
  if (prerelease.some((part) => !part || (/^\d+$/.test(part) && part.length > 1 && part.startsWith("0")))) {
    return null
  }

  return { major, minor, patch, prerelease }
}

export function normalizeDesktopVersion(raw: string | null | undefined): string | null {
  const parsed = parseDesktopSemver(raw)
  if (!parsed) return null
  const base = `${parsed.major}.${parsed.minor}.${parsed.patch}`
  return parsed.prerelease.length ? `${base}-${parsed.prerelease.join(".")}` : base
}

function comparePrereleaseIdentifier(left: string, right: string): number {
  const leftNumeric = /^\d+$/.test(left)
  const rightNumeric = /^\d+$/.test(right)
  if (leftNumeric && rightNumeric) return Number(left) - Number(right)
  if (leftNumeric) return -1
  if (rightNumeric) return 1
  return left.localeCompare(right, "en")
}

function comparePrerelease(left: string[], right: string[]): number {
  if (left.length === 0 && right.length === 0) return 0
  if (left.length === 0) return 1
  if (right.length === 0) return -1

  const len = Math.max(left.length, right.length)
  for (let i = 0; i < len; i += 1) {
    const l = left[i]
    const r = right[i]
    if (l === undefined) return -1
    if (r === undefined) return 1
    const cmp = comparePrereleaseIdentifier(l, r)
    if (cmp !== 0) return cmp
  }
  return 0
}

export function compareDesktopVersions(
  left: string | null | undefined,
  right: string | null | undefined,
): number | null {
  const a = parseDesktopSemver(left)
  const b = parseDesktopSemver(right)
  if (!a || !b) return null

  for (const key of ["major", "minor", "patch"] as const) {
    const diff = a[key] - b[key]
    if (diff !== 0) return diff
  }
  return comparePrerelease(a.prerelease, b.prerelease)
}

export function latestDesktopVersionFromManifest(
  manifest: DesktopUpdateManifest | null | undefined,
): string | null {
  return normalizeDesktopVersion(manifest?.semver) ?? normalizeDesktopVersion(manifest?.version)
}

function unknownErrorMessage(error: unknown): string {
  if (error instanceof Error && error.message.trim()) return error.message
  return String(error || "未知错误")
}

export function shouldShowDesktopUpdateNotice(
  result: Pick<DesktopUpdateCheckResult, "ok" | "updateAvailable" | "latestVersion">,
): boolean {
  return Boolean(result.ok && result.updateAvailable && result.latestVersion)
}

export function dispatchDesktopUpdateDialog(result: DesktopUpdateCheckResult): void {
  window.dispatchEvent(new CustomEvent<DesktopUpdateCheckResult>(DESKTOP_UPDATE_DIALOG_EVENT, { detail: result }))
}

export function buildDesktopUpdateCheckResult(
  fetchResult: DesktopUpdateManifestFetchResult,
  currentVersion = DESKTOP_VERSION,
): DesktopUpdateCheckResult {
  const current = normalizeDesktopVersion(currentVersion)
  const latest = latestDesktopVersionFromManifest(fetchResult.manifest)
  const sourceUrl = fetchResult.manifest?.sourceUrl

  if (!fetchResult.ok) {
    return {
      ...fetchResult,
      ok: false,
      updateAvailable: false,
      currentVersion: current ?? currentVersion,
      latestVersion: latest ?? undefined,
      downloadUrl: DESKTOP_UPDATE_DOWNLOAD_URL,
      sourceUrl,
      error: fetchResult.error ?? "桌面端更新检查失败",
    }
  }

  if (!current) {
    return {
      ...fetchResult,
      ok: false,
      updateAvailable: false,
      currentVersion,
      latestVersion: latest ?? undefined,
      downloadUrl: DESKTOP_UPDATE_DOWNLOAD_URL,
      sourceUrl,
      error: `当前桌面端版本无效：${currentVersion}`,
    }
  }

  if (!latest) {
    return {
      ...fetchResult,
      ok: false,
      updateAvailable: false,
      currentVersion: current,
      downloadUrl: DESKTOP_UPDATE_DOWNLOAD_URL,
      sourceUrl,
      error: "桌面端更新清单缺少有效版本号",
    }
  }

  const cmp = compareDesktopVersions(latest, current)
  if (cmp == null) {
    return {
      ...fetchResult,
      ok: false,
      updateAvailable: false,
      currentVersion: current,
      latestVersion: latest,
      downloadUrl: DESKTOP_UPDATE_DOWNLOAD_URL,
      sourceUrl,
      error: "桌面端版本号比较失败",
    }
  }

  return {
    ...fetchResult,
    ok: true,
    updateAvailable: cmp > 0,
    currentVersion: current,
    latestVersion: latest,
    downloadUrl: DESKTOP_UPDATE_DOWNLOAD_URL,
    sourceUrl,
    error: undefined,
  }
}

export function bridgeUnavailableDesktopUpdateResult(currentVersion = DESKTOP_VERSION): DesktopUpdateCheckResult {
  const current = normalizeDesktopVersion(currentVersion) ?? currentVersion
  return {
    ok: false,
    updateAvailable: false,
    currentVersion: current,
    manifestUrl: DESKTOP_UPDATE_MANIFEST_URL,
    downloadUrl: DESKTOP_UPDATE_DOWNLOAD_URL,
    checkedAtMs: Date.now(),
    error: "当前环境没有桌面端更新检查能力",
  }
}

export async function checkDesktopUpdate(currentVersion = DESKTOP_VERSION): Promise<DesktopUpdateCheckResult> {
  try {
    const fetchResult = await invoke<DesktopUpdateManifestFetchResult>("desktop_check_update")
    return buildDesktopUpdateCheckResult(fetchResult, currentVersion)
  } catch (error) {
    return {
      ...bridgeUnavailableDesktopUpdateResult(currentVersion),
      error: `桌面端更新检查失败：${unknownErrorMessage(error)}`,
    }
  }
}

export async function installDesktopUpdate(): Promise<DesktopInstallUpdateResult> {
  return invoke<DesktopInstallUpdateResult>("desktop_install_update")
}

File: src-tauri/src/commands/desktop_update.rs
use std::collections::BTreeMap;
use std::fs::{self, File};
use std::io::{Read, Write};
use std::path::{Path, PathBuf};
use std::sync::LazyLock;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use futures_util::StreamExt;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use tauri::{AppHandle, Emitter};

const DEFAULT_UPDATE_MANIFEST_URL: &str =
    "https://ai.fengchiyun.com/downloads/huanxingtradereview/latest.json";
const DESKTOP_UPDATE_TIMEOUT: Duration = Duration::from_secs(10);
const DESKTOP_INSTALL_TIMEOUT: Duration = Duration::from_secs(20 * 60);
pub const DESKTOP_UPDATE_PROGRESS_EVENT: &str = "desktop-update-progress";

static DESKTOP_UPDATE_HTTP_CLIENT: LazyLock<reqwest::Client> = LazyLock::new(|| {
    reqwest::Client::builder()
        .timeout(DESKTOP_UPDATE_TIMEOUT)
        .user_agent("trading-review-wiki-desktop-update-check")
        .build()
        .expect("valid desktop update HTTP client")
});

static DESKTOP_INSTALL_HTTP_CLIENT: LazyLock<reqwest::Client> = LazyLock::new(|| {
    reqwest::Client::builder()
        .timeout(DESKTOP_INSTALL_TIMEOUT)
        .user_agent("trading-review-wiki-desktop-update-install")
        .build()
        .expect("valid desktop update install HTTP client")
});

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DesktopUpdateAsset {
    pub label: Option<String>,
    pub platform: Option<String>,
    pub file_name: Option<String>,
    pub size: Option<u64>,
    pub sha256: Option<String>,
    pub url: Option<String>,
    pub versioned_url: Option<String>,
    pub source_url: Option<String>,
    pub baidu_pan_url: Option<String>,
    pub baidu_pan_code: Option<String>,
    pub quark_pan_url: Option<String>,
    pub quark_pan_code: Option<String>,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DesktopUpdateManifest {
    pub repository: Option<String>,
    pub version: Option<String>,
    pub semver: Option<String>,
    pub published_at: Option<String>,
    pub source_url: Option<String>,
    pub updated_at: Option<String>,
    pub assets: Option<BTreeMap<String, DesktopUpdateAsset>>,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DesktopUpdateManifestFetchResult {
    pub ok: bool,
    pub manifest_url: String,
    pub manifest: Option<DesktopUpdateManifest>,
    pub error: Option<String>,
    pub checked_at_ms: u64,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DesktopInstallUpdateResult {
    pub ok: bool,
    pub manifest_url: String,
    pub asset: Option<DesktopUpdateAsset>,
    pub file_path: Option<String>,
    pub bytes_downloaded: u64,
    pub bytes_total: Option<u64>,
    pub launched: bool,
    pub error: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DesktopInstallUpdateProgress {
    pub stage: &'static str,
    pub bytes_downloaded: u64,
    pub bytes_total: Option<u64>,
    pub percent: Option<u8>,
    pub file_name: Option<String>,
    pub message: Option<String>,
}

fn update_manifest_url() -> &'static str {
    option_env!("TRADING_REVIEW_UPDATE_MANIFEST_URL").unwrap_or(DEFAULT_UPDATE_MANIFEST_URL)
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis()
        .try_into()
        .unwrap_or(u64::MAX)
}

fn download_percent(bytes_downloaded: u64, bytes_total: Option<u64>) -> Option<u8> {
    let total = bytes_total?;
    if total == 0 {
        return None;
    }
    Some(((bytes_downloaded.saturating_mul(100) / total).min(100)) as u8)
}

fn emit_install_progress(
    app: Option<&AppHandle>,
    stage: &'static str,
    bytes_downloaded: u64,
    bytes_total: Option<u64>,
    file_name: Option<String>,
    message: Option<&str>,
) {
    let Some(app) = app else {
        return;
    };
    let _ = app.emit(
        DESKTOP_UPDATE_PROGRESS_EVENT,
        DesktopInstallUpdateProgress {
            stage,
            bytes_downloaded,
            bytes_total,
            percent: download_percent(bytes_downloaded, bytes_total),
            file_name,
            message: message.map(str::to_string),
        },
    );
}

async fn fetch_desktop_update_manifest_from(
    client: &reqwest::Client,
    manifest_url: &str,
) -> DesktopUpdateManifestFetchResult {
    let checked_at_ms = now_ms();
    let response = match client
        .get(manifest_url)
        .header(reqwest::header::ACCEPT, "application/json")
        .send()
        .await
    {
        Ok(response) => response,
        Err(error) => {
            return DesktopUpdateManifestFetchResult {
                ok: false,
                manifest_url: manifest_url.to_string(),
                manifest: None,
                error: Some(format!("桌面端更新清单请求失败：{error}")),
                checked_at_ms,
            }
        }
    };

    let status = response.status();
    if !status.is_success() {
        return DesktopUpdateManifestFetchResult {
            ok: false,
            manifest_url: manifest_url.to_string(),
            manifest: None,
            error: Some(format!("桌面端更新清单返回 HTTP {}", status.as_u16())),
            checked_at_ms,
        };
    }

    match response.json::<DesktopUpdateManifest>().await {
        Ok(manifest) => DesktopUpdateManifestFetchResult {
            ok: true,
            manifest_url: manifest_url.to_string(),
            manifest: Some(manifest),
            error: None,
            checked_at_ms,
        },
        Err(error) => DesktopUpdateManifestFetchResult {
            ok: false,
            manifest_url: manifest_url.to_string(),
            manifest: None,
            error: Some(format!("桌面端更新清单解析失败：{error}")),
            checked_at_ms,
        },
    }
}

#[tauri::command]
pub async fn desktop_check_update() -> DesktopUpdateManifestFetchResult {
    fetch_desktop_update_manifest_from(&DESKTOP_UPDATE_HTTP_CLIENT, update_manifest_url()).await
}

#[tauri::command]
pub async fn desktop_install_update(app: AppHandle) -> DesktopInstallUpdateResult {
    install_desktop_update_from(
        &DESKTOP_UPDATE_HTTP_CLIENT,
        &DESKTOP_INSTALL_HTTP_CLIENT,
        update_manifest_url(),
        Some(&app),
    )
    .await
}

async fn install_desktop_update_from(
    manifest_client: &reqwest::Client,
    download_client: &reqwest::Client,
    manifest_url: &str,
    app: Option<&AppHandle>,
) -> DesktopInstallUpdateResult {
    emit_install_progress(app, "starting", 0, None, None, Some("正在读取更新清单"));
    let fetched = fetch_desktop_update_manifest_from(manifest_client, manifest_url).await;
    if !fetched.ok {
        if let Some(error) = fetched.error.as_deref() {
            emit_install_progress(app, "error", 0, None, None, Some(error));
        }
        return DesktopInstallUpdateResult {
            ok: false,
            manifest_url: fetched.manifest_url,
            asset: None,
            file_path: None,
            bytes_downloaded: 0,
            bytes_total: None,
            launched: false,
            error: fetched.error,
        };
    }

    let Some(manifest) = fetched.manifest else {
        return install_error(app, manifest_url, None, 0, None, "桌面端更新清单为空");
    };
    let Some(asset) = select_platform_asset(&manifest) else {
        return install_error(
            app,
            manifest_url,
            None,
            0,
            None,
            "更新清单中没有当前系统可用的安装包",
        );
    };
    let Some(download_url) = asset_download_url(&asset) else {
        return install_error(
            app,
            manifest_url,
            Some(asset),
            0,
            None,
            "安装包缺少下载地址",
        );
    };

    let raw_file_name = asset
        .file_name
        .clone()
        .or_else(|| url_file_name(&download_url))
        .unwrap_or_else(|| default_installer_file_name().to_string());
    let file_name = safe_file_name(&raw_file_name);
    let dir = desktop_update_download_dir();
    if let Err(err) = fs::create_dir_all(&dir) {
        return install_error(
            app,
            manifest_url,
            Some(asset),
            0,
            None,
            &format!("创建下载目录失败：{err}"),
        );
    }
    let file_path = dir.join(file_name);

    let downloaded = match download_installer(download_client, &download_url, &file_path, app).await
    {
        Ok(downloaded) => downloaded,
        Err(err) => {
            return install_error(
                app,
                manifest_url,
                Some(asset),
                err.bytes_downloaded,
                err.bytes_total,
                &err.message,
            )
        }
    };

    let file_name_for_progress = file_path
        .file_name()
        .and_then(|name| name.to_str())
        .map(str::to_string);
    emit_install_progress(
        app,
        "verifying",
        downloaded.bytes_downloaded,
        downloaded.bytes_total,
        file_name_for_progress.clone(),
        Some("正在校验安装包"),
    );

    if let Some(expected) = asset
        .sha256
        .as_deref()
        .map(str::trim)
        .filter(|v| !v.is_empty())
        .map(str::to_string)
    {
        match file_sha256(&file_path) {
            Ok(actual) if actual.eq_ignore_ascii_case(&expected) => {}
            Ok(actual) => {
                return install_error(
                    app,
                    manifest_url,
                    Some(asset),
                    downloaded.bytes_downloaded,
                    downloaded.bytes_total,
                    &format!("安装包 SHA-256 校验失败：期望 {expected}，实际 {actual}"),
                )
            }
            Err(err) => {
                return install_error(
                    app,
                    manifest_url,
                    Some(asset),
                    downloaded.bytes_downloaded,
                    downloaded.bytes_total,
                    &format!("读取安装包校验失败：{err}"),
                )
            }
        }
    }

    emit_install_progress(
        app,
        "launching",
        downloaded.bytes_downloaded,
        downloaded.bytes_total,
        file_name_for_progress.clone(),
        Some("正在打开安装包"),
    );
    if let Err(err) = open::that(&file_path) {
        return install_error(
            app,
            manifest_url,
            Some(asset),
            downloaded.bytes_downloaded,
            downloaded.bytes_total,
            &format!("启动安装包失败：{err}"),
        );
    }

    emit_install_progress(
        app,
        "complete",
        downloaded.bytes_downloaded,
        downloaded.bytes_total,
        file_name_for_progress,
        Some("安装包已打开"),
    );

    DesktopInstallUpdateResult {
        ok: true,
        manifest_url: manifest_url.to_string(),
        asset: Some(asset),
        file_path: Some(file_path.to_string_lossy().to_string()),
        bytes_downloaded: downloaded.bytes_downloaded,
        bytes_total: downloaded.bytes_total,
        launched: true,
        error: None,
    }
}

fn install_error(
    app: Option<&AppHandle>,
    manifest_url: &str,
    asset: Option<DesktopUpdateAsset>,
    bytes_downloaded: u64,
    bytes_total: Option<u64>,
    message: &str,
) -> DesktopInstallUpdateResult {
    let file_name = asset.as_ref().and_then(|asset| asset.file_name.clone());
    emit_install_progress(
        app,
        "error",
        bytes_downloaded,
        bytes_total,
        file_name,
        Some(message),
    );
    DesktopInstallUpdateResult {
        ok: false,
        manifest_url: manifest_url.to_string(),
        asset,
        file_path: None,
        bytes_downloaded,
        bytes_total,
        launched: false,
        error: Some(message.to_string()),
    }
}

struct DownloadedInstaller {
    bytes_downloaded: u64,
    bytes_total: Option<u64>,
}

struct DownloadInstallError {
    message: String,
    bytes_downloaded: u64,
    bytes_total: Option<u64>,
}

async fn download_installer(
    client: &reqwest::Client,
    url: &str,
    target: &Path,
    app: Option<&AppHandle>,
) -> Result<DownloadedInstaller, DownloadInstallError> {
    let response = client
        .get(url)
        .send()
        .await
        .map_err(|err| DownloadInstallError {
            message: format!("安装包下载请求失败：{err}"),
            bytes_downloaded: 0,
            bytes_total: None,
        })?;
    let status = response.status();
    let bytes_total = response.content_length();
    if !status.is_success() {
        return Err(DownloadInstallError {
            message: format!("安装包下载返回 HTTP {}", status.as_u16()),
            bytes_downloaded: 0,
            bytes_total,
        });
    }

    let file_name = target
        .file_name()
        .and_then(|name| name.to_str())
        .map(str::to_string);
    emit_install_progress(
        app,
        "downloading",
        0,
        bytes_total,
        file_name.clone(),
        Some("正在下载安装包"),
    );

    let temp_path = target.with_extension(format!(
        "{}download",
        target
            .extension()
            .and_then(|ext| ext.to_str())
            .map(|ext| format!("{ext}."))
            .unwrap_or_default()
    ));
    let mut file = File::create(&temp_path).map_err(|err| DownloadInstallError {
        message: format!("创建安装包文件失败：{err}"),
        bytes_downloaded: 0,
        bytes_total,
    })?;
    let mut bytes_downloaded = 0_u64;
    let mut stream = response.bytes_stream();
    while let Some(chunk) = stream.next().await {
        let chunk = chunk.map_err(|err| DownloadInstallError {
            message: format!("安装包下载中断：{err}"),
            bytes_downloaded,
            bytes_total,
        })?;
        file.write_all(&chunk).map_err(|err| DownloadInstallError {
            message: format!("写入安装包失败：{err}"),
            bytes_downloaded,
            bytes_total,
        })?;
        bytes_downloaded = bytes_downloaded.saturating_add(chunk.len() as u64);
        emit_install_progress(
            app,
            "downloading",
            bytes_downloaded,
            bytes_total,
            file_name.clone(),
            Some("正在下载安装包"),
        );
    }
    file.flush().map_err(|err| DownloadInstallError {
        message: format!("刷新安装包文件失败：{err}"),
        bytes_downloaded,
        bytes_total,
    })?;
    drop(file);
    fs::rename(&temp_path, target).map_err(|err| DownloadInstallError {
        message: format!("保存安装包失败：{err}"),
        bytes_downloaded,
        bytes_total,
    })?;
    Ok(DownloadedInstaller {
        bytes_downloaded,
        bytes_total,
    })
}

fn desktop_update_download_dir() -> PathBuf {
    dirs::download_dir()
        .or_else(dirs::data_dir)
        .unwrap_or_else(std::env::temp_dir)
        .join("HuanxingTradeReview Updates")
}

fn safe_file_name(input: &str) -> String {
    let cleaned = input
        .chars()
        .map(|ch| match ch {
            '/' | '\\' | ':' | '*' | '?' | '"' | '<' | '>' | '|' => '_',
            ch if ch.is_control() => '_',
            ch => ch,
        })
        .collect::<String>();
    let trimmed = cleaned.trim().trim_matches('.').trim();
    if trimmed.is_empty() {
        default_installer_file_name().to_string()
    } else {
        trimmed.to_string()
    }
}

fn default_installer_file_name() -> &'static str {
    if cfg!(target_os = "windows") {
        "desktop-update-setup.exe"
    } else if cfg!(target_os = "macos") {
        "desktop-update.dmg"
    } else {
        "desktop-update"
    }
}

fn url_file_name(raw: &str) -> Option<String> {
    let url = url::Url::parse(raw).ok()?;
    url.path_segments()
        .and_then(|segments| segments.filter(|s| !s.is_empty()).last())
        .map(urlencoding::decode)
        .and_then(Result::ok)
        .map(|cow| cow.into_owned())
}

fn asset_download_url(asset: &DesktopUpdateAsset) -> Option<String> {
    [
        asset.versioned_url.as_deref(),
        asset.url.as_deref(),
        asset.source_url.as_deref(),
    ]
    .into_iter()
    .flatten()
    .map(str::trim)
    .find(|url| url.starts_with("https://") || url.starts_with("http://"))
    .map(str::to_string)
}

fn select_platform_asset(manifest: &DesktopUpdateManifest) -> Option<DesktopUpdateAsset> {
    let assets = manifest.assets.as_ref()?;
    assets
        .iter()
        .filter_map(|(key, asset)| {
            let score = platform_asset_score(key, asset);
            (score > 0).then_some((score, asset.clone()))
        })
        .max_by_key(|(score, _)| *score)
        .map(|(_, asset)| asset)
}

fn platform_asset_score(key: &str, asset: &DesktopUpdateAsset) -> i32 {
    let haystack = [
        key.to_ascii_lowercase(),
        asset
            .platform
            .clone()
            .unwrap_or_default()
            .to_ascii_lowercase(),
        asset
            .file_name
            .clone()
            .unwrap_or_default()
            .to_ascii_lowercase(),
        asset.label.clone().unwrap_or_default().to_ascii_lowercase(),
    ]
    .join(" ");

    if cfg!(target_os = "windows") {
        if haystack.contains("windows") || haystack.contains("win32") {
            return 100 + if haystack.contains("x64") { 10 } else { 0 };
        }
        if haystack.ends_with(".exe") || haystack.contains("setup.exe") {
            return 70;
        }
    } else if cfg!(target_os = "macos") {
        let arch_match = if cfg!(target_arch = "aarch64") {
            haystack.contains("arm64") || haystack.contains("aarch64") || haystack.contains("apple")
        } else {
            haystack.contains("x64") || haystack.contains("x86_64") || haystack.contains("intel")
        };
        if (haystack.contains("macos") || haystack.contains("darwin")) && arch_match {
            return 110;
        }
        if haystack.contains("macos") || haystack.contains("darwin") || haystack.ends_with(".dmg") {
            return 80;
        }
    }
    0
}

fn file_sha256(path: &Path) -> std::io::Result<String> {
    let mut file = File::open(path)?;
    let mut hasher = Sha256::new();
    let mut buf = [0_u8; 64 * 1024];
    loop {
        let n = file.read(&mut buf)?;
        if n == 0 {
            break;
        }
        hasher.update(&buf[..n]);
    }
    Ok(format!("{:x}", hasher.finalize()))
}
